### PR TITLE
Image comparison

### PR DESCRIPTION
**Pull request by tyBouch**
Ready for code review.
Added background color fetching to ensure that with image comparison we're aware of the current background color. This also adds support to getting background color for any tests requiring it.

### CODE REVIEW

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java [hunk 1]
```diff
@@ -179,4 +181,41 @@ public void sendSpecialKey() {
...
+		Color startingColor = Color.white;
+		String hexOfStartingColor = "#"+Integer.toHexString(startingColor.getRGB()).substring(2);
+		Color endingColor = Color.blue;
+		String hexOfEndingColor = "#"+Integer.toHexString(endingColor.getRGB()).substring(2);
+		
+		BaseSteps.navigateToPage("TextboxPage");
+		js.executeScript("arguments[0].style.backgroundColor=arguments[1];", getElement("Body").element(), hexOfStartingColor);
+		Assert.assertEquals(startingColor, getElement("Car Checkbox").getBackgroundColor());
+		js.executeScript("arguments[0].style.backgroundColor=arguments[1];", getElement("Body").element(), hexOfEndingColor);
+		Assert.assertEquals(endingColor, getElement("Car Checkbox").getBackgroundColor());
```
Comment (dougnoel) [inline src/test/java/com/dougnoel/sentinel/elements/ElementTests.java:197]: There are two assertions in this unit test. These should be two separate tests.

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java [hunk 2]
```diff
@@ -179,4 +181,41 @@ public void sendSpecialKey() {
...
+		Color startingColor = Color.black;
+		String hexOfStartingColor = "#"+Integer.toHexString(startingColor.getRGB()).substring(2);
+		Color endingColor = Color.red;
+		String hexOfEndingColor = "#"+Integer.toHexString(endingColor.getRGB()).substring(2);
+		
+		BaseSteps.navigateToPage("TextboxPage");
+		js.executeScript("arguments[0].style.backgroundColor=arguments[1];", getElement("Car Checkbox").element(), hexOfStartingColor);
+		Assert.assertEquals(startingColor, getElement("Car Checkbox").getBackgroundColor());
+		js.executeScript("arguments[0].style.backgroundColor=arguments[1];", getElement("Car Checkbox").element(), hexOfEndingColor);
+		Assert.assertEquals(endingColor, getElement("Car Checkbox").getBackgroundColor());
```
Comment (dougnoel) [inline src/test/java/com/dougnoel/sentinel/elements/ElementTests.java:213]: There are two assertions in this unit test. These should be two separate tests.

File: src/test/java/com/dougnoel/sentinel/pages/PageManagerTests.java [hunk 3]
```diff
@@ -31,4 +35,26 @@ public void getWindowHandleInSameWindow() throws InterruptedException {
 			//suppress
 		}
 	}
+	
+	@Test
+	public void checkBackgroundColorOfPage() {
```
Comment (dougnoel) [inline src/test/java/com/dougnoel/sentinel/pages/PageManagerTests.java:40]: Can these tests cover the same code as the ones in Element tests? What's the minimum amount of tests we can do to get code coverage? And if the answer is no, can you refactor the code so these two objects use the same code?

**Review by dougnoel — CHANGES_REQUESTED** [review]

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 4]
```diff
@@ -695,6 +703,60 @@ public Element hover() {
...
+		byte[] imageAsByteArray = element().getScreenshotAs(OutputType.BYTES);
+		ByteArrayInputStream imageByteStream = new ByteArrayInputStream(imageAsByteArray);
+
+		BufferedImage elementImage = ImageIO.read(imageByteStream);
+		BufferedImage scaledImage = (BufferedImage) elementImage.getScaledInstance(1, 1, Image.SCALE_REPLICATE);
+		var rgb = scaledImage.getRGB(0, 0);
+		return new Color(rgb);
+	}
+	
+	public File getScreenshot() {
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java:718]: Added a header comment of the functionality.

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 5]
```diff
@@ -695,6 +703,60 @@ public Element hover() {
 	public String getTooltipText() {
 		hover();
 		return driver().findElement(By.xpath("//*[contains(text(),'')]")).getText();
-	}	
+	}
+	
+	public Color getMostPrevalentColor() throws IOException {
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java:708]: Added a header comment

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 6]
```diff
@@ -695,6 +703,60 @@ public Element hover() {
...
+	 * 
+	 * @return The background color of the element converted to a java color object.
+	 */
+	public Color getBackgroundColor()
+	{  
+		return ascendTransparentColorElements(element());
+	}
 	
+	/**
+	 * This method is used to ascend through elements with transparency to find a parent with a non transparent background color
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java:733]: I reworded it to not go into implementation and instead that it returns the color present behind the element. The user shouldn't need to worry about it's ascension method.

File: src/main/java/com/dougnoel/sentinel/pages/PageManager.java [hunk 7]
```diff
@@ -348,4 +352,31 @@ private static boolean isPageLoaded() {
...
+    		//If no colors are found, the page background defaults to white
+    		if(bgColor.equals(transparency)) {
+    			return Color.white;
+    		}
+    		else {
+    			return bgColor;
+    		}
+    	}
+		catch(NoSuchElementException e) {
+			//If there is no such body element, the page background defaults to white
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/pages/PageManager.java:378]: Done so

File: src/main/java/com/dougnoel/sentinel/pages/PageManager.java [hunk 8]
```diff
@@ -348,4 +352,31 @@ private static boolean isPageLoaded() {
...
+    		Color bgColor = org.openqa.selenium.support.Color.fromString(color).getColor();
+    		Color transparency = org.openqa.selenium.support.Colors.TRANSPARENT.getColorValue().getColor();
+    		
+    		//If no colors are found, the page background defaults to white
+    		if(bgColor.equals(transparency)) {
+    			return Color.white;
+    		}
+    		else {
+    			return bgColor;
+    		}
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/pages/PageManager.java:375]: Made the change!

File: src/main/java/com/dougnoel/sentinel/steps/BaseSteps.java [hunk 9]
```diff
@@ -1,7 +1,6 @@
 package com.dougnoel.sentinel.steps;
 
 import static com.dougnoel.sentinel.elements.ElementFunctions.getElement;
-
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/steps/BaseSteps.java:4]: Done!

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 10]
```diff
@@ -695,6 +703,69 @@ public Element hover() {
...
+			if(parentElement != null) {
+				return(ascendTransparentColorElements(parentElement));
+			}
+			else {
+				return Color.white;
+			}
+		} else {
+			return currentColor;
+		}
+	}
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/elements/Element.java:770]: The logic of this is a bit convoluted. Look up Big O notation for an in-depth discussion of the math, but basically for a self-recursive method we want as little branching and if statements as possible. We can reduce that by getting rid of branches. First we can get rid of the last else.
```
		Color currentColor = org.openqa.selenium.support.Color.fromString(element.getCssValue("background-color")).getColor();
		WebElement parentElement = null;

		if(currentColor.equals(transparent)) {
			try {
				parentElement = element.findElement(By.xpath("./.."));
			} catch(NoSuchElementException e) {
				//Suppress error when there are no more parents
			}
			
			if(parentElement != null) {
				return(ascendTransparentColorElements(parentElement));
			}
			else {
				return Color.white;
			}
		}
		return currentColor;
```

Now if we look at the logic in your recursion, we stop when there are no more parents - when we get a NoSuchElementException at that point we test farther below and return white. But as soon as that exception is thrown, we know we can return that color. By doing it there, we remove the need for another branch. It also solves the problem that you should never suppress exceptions. They should always be handled or thrown.

```
		Color currentColor = org.openqa.selenium.support.Color.fromString(element.getCssValue("background-color")).getColor();
		WebElement parentElement = null;

		if(currentColor.equals(transparent)) {
			try {
				parentElement = element.findElement(By.xpath("./.."));
			} catch(NoSuchElementException e) {
				return Color.white;
			}

			if(parentElement != null) {
				return(ascendTransparentColorElements(parentElement));
			}
		}
		return currentColor;
```

Our cognitive complexity is reducing, making the code easier to read, and that is also making the code execute faster because less code is being compiled in our recursive function. But we can reduce it further. We are doing a null test because we want to make sure parentElement is valid. But, because we are returning if it isn't, we no longer need that chack. We can take that branch out entirely. While we are at it, the return statement doesn't need parentheses around it so we should fix that.

```
		Color currentColor = org.openqa.selenium.support.Color.fromString(element.getCssValue("background-color")).getColor();
		WebElement parentElement = null;

		if(currentColor.equals(transparent)) {
			try {
				parentElement = element.findElement(By.xpath("./.."));
			} catch(NoSuchElementException e) {
				return Color.white;
			}

			return ascendTransparentColorElements(parentElement);
		}
		return currentColor;
```

Now let's talk about memory usage. parentElement is only used on those two lines of code. We can collapse them into one line to avoid using the variable at all and save some memory. While we are at it I'm going to rename the method to match the comment I made on line 749. If the find method execution fails it will throw the exception and the method will never be called recursively, instead, it will return white.

```
		Color currentColor = org.openqa.selenium.support.Color.fromString(element.getCssValue("background-color")).getColor();

		if(currentColor.equals(transparent)) {
			try {
				return getBackgroundColor(element.findElement(By.xpath("./.."));
			} catch(NoSuchElementException e) {
				return Color.white;
			}
		}
		return currentColor;
```

We've gone from 19 lines of code to 10. This is pretty good, but we can still optimize the speed. We still have an implicit if-else and so we can increase execution speed by returning the positive value first, so that we do not have to evaluate the rest of the code. As soon as we know what we are going to return, we return it.

```suggestion
		Color currentColor = org.openqa.selenium.support.Color.fromString(element.getCssValue("background-color")).getColor();

		if(!currentColor.equals(transparent))
			return currentColor;
		try {
			return getBackgroundColor(element.findElement(By.xpath("./.."));
		} catch(NoSuchElementException e) {
			return Color.white;
		}
```

We've now gotten it done to 9 lines from 19, eliminated a variable, made the execution faster, and gotten rid of a number of branching statements. (This also makes the unit test easier.)

**Review by dougnoel — CHANGES_REQUESTED** [review]
I will have more but I wanted to get you started.

File: src/main/java/com/dougnoel/sentinel/configurations/Configuration.java [hunk 11]
```diff
@@ -405,6 +405,19 @@ public static String operatingSystem() {
...
+     */
+    public static String imageDirectory() {
+    	String imageDirectory = Configuration.toString("imageDirectory");
+        if(imageDirectory == null) {
+    		imageDirectory = "logs/images";
+    	}
+        
+        return imageDirectory;
+    }
+
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/configurations/Configuration.java:420]: Move this method to the FileManager. That's the object that needs to know what the path is and handle it.

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java [hunk 12]
```diff
@@ -179,4 +181,33 @@ public void sendSpecialKey() {
...
+	@Test
+	public void checkParentColorsOfTransparentElement() {
+		JavascriptExecutor js = (JavascriptExecutor) WebDriverFactory.getWebDriver(); 
+		
+		Color endingColor = Color.blue;
+		String hexOfEndingColor = "#"+Integer.toHexString(endingColor.getRGB()).substring(2);
+		
+		BaseSteps.navigateToPage("TextboxPage");
+		js.executeScript("arguments[0].style.backgroundColor=arguments[1];", getElement("Body").element(), hexOfEndingColor);
+		Assert.assertEquals(endingColor, getElement("Car Checkbox").getBackgroundColor());
```
Comment (dougnoel) [inline src/test/java/com/dougnoel/sentinel/elements/ElementTests.java:193]: Change the sndingColorVariable to something like colorBlue and hexOfEndingColor to whatever it is, the blueValue? Variable names should be descriptive, not enigmatic.

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java [hunk 13]
```diff
@@ -179,4 +181,33 @@ public void sendSpecialKey() {
...
+	@Test
+	public void checkElementBackgroundColor() {
+		JavascriptExecutor js = (JavascriptExecutor) WebDriverFactory.getWebDriver(); 
+		
+		Color endingColor = Color.red;
+		String hexOfEndingColor = "#"+Integer.toHexString(endingColor.getRGB()).substring(2);
+		
+		BaseSteps.navigateToPage("TextboxPage");
+		js.executeScript("arguments[0].style.backgroundColor=arguments[1];", getElement("Car Checkbox").element(), hexOfEndingColor);
+		Assert.assertEquals(endingColor, getElement("Car Checkbox").getBackgroundColor());
```
Comment (dougnoel) [inline src/test/java/com/dougnoel/sentinel/elements/ElementTests.java:205]: Same note as above.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java [hunk 14]
```diff
@@ -98,4 +104,60 @@ public static String getClassPath(String className) {
...
+		
+		File destinationFile = new File(Configuration.imageDirectory() + "/" + subDirectory + imageFileName);
+    	FileUtils.copyFile(imageFile, destinationFile);
+	}
+	
+	/**
+	 * Saves an image of type BufferedImage to the given sub-directory of the configured, or default, image path.
+	 * 
+	 * @param subDirectory the sub-directory within the configured or default image path.
+	 * If the subDirectory is null, the root image directory will be used.
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:130]: This line should go in the description above.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java [hunk 15]
```diff
@@ -98,4 +104,60 @@ public static String getClassPath(String className) {
...
+	 * @param imageFile the BufferedImage of the image to save.
+	 */
+	public static void saveImage(String subDirectory, String imageFileName, BufferedImage imageFile) throws IOException {
+		if(subDirectory == null) 
+			subDirectory = "";
+		else
+			subDirectory += "/";
+		
+		File destinationFile = new File(Configuration.imageDirectory() + "/" + subDirectory + imageFileName);
+		//ImageIO will not create directories. We have to make them ourselves before running the write command.
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:141]: Unnecessary comment.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java [hunk 16]
```diff
@@ -98,4 +104,60 @@ public static String getClassPath(String className) {
...
+	 * @param subDirectory the sub-directory within the configured or default image path.
+	 * If the sub-directory is null, the root image directory will be used.
+	 * @param imageFileName the file name of the image to save.
+	 * @param imageFile the File of the image to save.
+	 */
+	public static void saveImage(String subDirectory, String imageFileName, File imageFile) throws IOException {
+		if(subDirectory == null) 
+			subDirectory = "";
+		else
+			subDirectory += "/";
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:120]: You repeat this code 3 times in three different methods. Move it into a separate method.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java [hunk 17]
```diff
@@ -98,4 +104,60 @@ public static String getClassPath(String className) {
...
+		//ImageIO will not create directories. We have to make them ourselves before running the write command.
+		FileUtils.forceMkdir(destinationFile.getParentFile());
+    	ImageIO.write(imageFile, "png", destinationFile);
+	}
+
+	/**
+	 * Reads an image from the sub-directory of the default or configured image directory.
+	 * 
+	 * @param sourceImageSubDirectory the sub-directory within the configured or default image path to read the file from.
+	 * If the sub-directory is null, the root image directory will be used.
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:150]: This line should go in the description above.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java [hunk 18]
```diff
@@ -98,4 +104,60 @@ public static String getClassPath(String className) {
 			return null;
 		}
 	}
+	
+	/**
+	 * Saves an image of type File to the given sub-directory of the configured, or default, image path.
+	 * 
+	 * @param subDirectory the sub-directory within the configured or default image path.
+	 * If the sub-directory is null, the root image directory will be used.
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:112]: This line should go in the description above.

File: src/main/java/com/dougnoel/sentinel/pages/PageManager.java [hunk 19]
```diff
@@ -3,16 +3,17 @@
...
-import org.openqa.selenium.NoSuchFrameException;
-import org.openqa.selenium.NoSuchWindowException;
-import org.openqa.selenium.NotFoundException;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.openqa.selenium.By;
 import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.NoSuchFrameException;
+import org.openqa.selenium.NoSuchWindowException;
+import org.openqa.selenium.NotFoundException;
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/pages/PageManager.java:12]: Let's just revert this file to what it was before the changes were made and leave it untouched.

**Review by dougnoel — CHANGES_REQUESTED** [review]
Finished my review.

File: src/main/java/com/dougnoel/sentinel/steps/ImageSteps.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    	
+    	//Determine if we're taking a screenshot of an element or the whole page.
+    	if(!elementName.toLowerCase().contentEquals("page")) {
+    		screenshotFile = getElement(elementName).getScreenshot();
+    	}
+    	else {
+        	TakesScreenshot pageScreenshotTool =((TakesScreenshot) WebDriverFactory.getWebDriver());
+        	screenshotFile = pageScreenshotTool.getScreenshotAs(OutputType.FILE);
+    	}
+    	File destinationFile = new File("logs/expected/" + imageFileName);
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/steps/ImageSteps.java]: Done

File: src/main/java/com/dougnoel/sentinel/steps/ImageVerificationSteps.java [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+	@Then("^I verify (?:the|an?) (.*?) (do(?:es)? not )?match(?:es)? the (?:expected|original) image$")
+    public static void verifyImageNotMatch(String elementName, String assertion) throws IOException {
+        boolean negate = !StringUtils.isEmpty(assertion);
+        var expectedResult = SentinelStringUtils.format("Expected {} to {}match its previous state visually.",
+                elementName, (negate ? "not " : ""));
+        
+		ImageComparisonResult comparisonResult = compareImages(elementName);
+		
+		String imageFileName = scenario.getName()+ "_" + elementName + "_" + PageManager.getPage().getName() + ".png";
+		File resultDestination = new File( "logs/diff/" + imageFileName );
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/steps/ImageVerificationSteps.java]: Done

File: src/main/java/com/dougnoel/sentinel/steps/ImageVerificationSteps.java [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+    		TakesScreenshot pageScreenshotTool =((TakesScreenshot) WebDriverFactory.getWebDriver());
+    		screenshotFile = pageScreenshotTool.getScreenshotAs(OutputType.FILE);
+    		backgroundColor = PageManager.getPageBackgroundColor();
+    	}
+    	
+    	File destinationFile = new File("logs/actual/" + imageFileName);
+    	FileUtils.copyFile(screenshotFile, destinationFile);
+
+    	//load images to be compared:
+        BufferedImage expectedImage = ImageComparisonUtil.readImageFromResources("logs/expected/" + imageFileName);
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/steps/ImageVerificationSteps.java]: Done

File: src/main/java/com/dougnoel/sentinel/steps/ImageVerificationSteps.java [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+        BufferedImage actualImage = ImageComparisonUtil.readImageFromResources("logs/actual/" + imageFileName);
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/steps/ImageVerificationSteps.java]: Done

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java (hunk 1 above)
Comment (tyBouch) [reply src/test/java/com/dougnoel/sentinel/elements/ElementTests.java]: Done

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java (hunk 2 above)
Comment (tyBouch) [reply src/test/java/com/dougnoel/sentinel/elements/ElementTests.java]: Done

File: src/test/java/com/dougnoel/sentinel/pages/PageManagerTests.java [hunk 24]
```diff
@@ -31,4 +35,26 @@ public void getWindowHandleInSameWindow() throws InterruptedException {
...
+		Color startingColor = Color.white;
+		String hexOfStartingColor = "#"+Integer.toHexString(startingColor.getRGB()).substring(2);
+		Color endingColor = Color.blue;
+		String hexOfEndingColor = "#"+Integer.toHexString(endingColor.getRGB()).substring(2);
+		
+		BaseSteps.navigateToPage("Add Remove Elements Page");
+		js.executeScript("$('body').css('background-color', arguments[0]);", hexOfStartingColor);
+		Assert.assertEquals(startingColor, PageManager.getPageBackgroundColor());
+		js.executeScript("$('body').css('background-color', arguments[0]);", hexOfEndingColor);
+		Assert.assertEquals(endingColor, PageManager.getPageBackgroundColor());
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/test/java/com/dougnoel/sentinel/pages/PageManagerTests.java]: Done

File: src/test/java/com/dougnoel/sentinel/pages/PageManagerTests.java (hunk 3 above)
Comment (tyBouch) [reply src/test/java/com/dougnoel/sentinel/pages/PageManagerTests.java]: Done

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 25]
```diff
@@ -695,6 +703,69 @@ public Element hover() {
...
+	 */
+	public File getScreenshot() {
+		return element().getScreenshotAs(OutputType.FILE);
+	}
+	
+	/**
+	 * This method is used to get the background color of the element, or it's inherited background color.
+	 * 
+	 * @return The background color of the element converted to a java color object.
+	 */
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java:736]: As the comments are already written mvn javadoc:javadoc would not have the intended effect to my understanding. I did what I could to follow what I understood of the javadocs section of the readme.

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 26]
```diff
@@ -695,6 +703,69 @@ public Element hover() {
...
+	{  
+		return ascendTransparentColorElements(element());
+	}
 	
+	/**
+	 * Returns the color of the background behind the element.
+	 * 
+	 * @param element The web element to use as the root for checking for background colors.
+	 * 
+	 * @return The background color of the first parent with one or white if no inherited background color is found.
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java]: Understood.

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 27]
```diff
@@ -695,6 +703,69 @@ public Element hover() {
...
+	 */
+	private Color ascendTransparentColorElements(WebElement element) {
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java]: Modified the documentation javadocs as well in order to get these up to date with eachother.

File: src/main/java/com/dougnoel/sentinel/elements/Element.java [hunk 28]
```diff
@@ -695,6 +703,69 @@ public Element hover() {
...
+		Color transparent = org.openqa.selenium.support.Colors.TRANSPARENT.getColorValue().getColor();
```
[the remark this answers is not in the document]
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/elements/Element.java]: Understood, and will do this moving forward. It was indeed a constant.

File: src/main/java/com/dougnoel/sentinel/configurations/Configuration.java (hunk 11 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/configurations/Configuration.java:420]: I moved it and updated the javadocs comments. Changed the calls to it to reflect this change.

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java (hunk 12 above)
Comment (tyBouch) [reply src/test/java/com/dougnoel/sentinel/elements/ElementTests.java:193]: Understood. Partially a relic of the original assertions when two were present. I've modified it accordingly and made the variable names more understandable

File: src/test/java/com/dougnoel/sentinel/elements/ElementTests.java (hunk 13 above)
Comment (tyBouch) [reply src/test/java/com/dougnoel/sentinel/elements/ElementTests.java:205]: Done

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java (hunk 15 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:141]: Wasn't sure if the code was understandable compared to the FileUtils.copyFile call used in the other saveImage. Removed.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java (hunk 14 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:130]: I rewrote the javadoc comments to put the null optional information higher up in a few places in FileManager

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java (hunk 16 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:120]: I consolidated this and a bunch of other reused code to createImagePath which creates an File to use for saving images. There's only one method that needs a string, and getAbsolutePath solves it there.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java (hunk 17 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:150]: Done

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java (hunk 18 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:112]: Done

File: src/main/java/com/dougnoel/sentinel/pages/PageManager.java (hunk 19 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/pages/PageManager.java:12]: Done.

**Comment by tyBouch** [comment]
Need to lookup merging later in git bash.

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java [hunk 29]
```diff
@@ -98,4 +104,78 @@ public static String getClassPath(String className) {
...
+	* @param subDirectory String the sub-directory to use
+	* @param fileName String the file name of the image
+	* 
+	* @return File the constructed image File from directories and filename
+	*/
+	private static File constructImagePath(String subDirectory, String fileName) {
+		if(subDirectory == null) 
+			subDirectory = "";
+		else
+			subDirectory += "/";
```
Comment (dougnoel) [inline src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:177]: 3 tests are failing on my Mac. I think it's related to this. We need to look at it together,

**Review by dougnoel — CHANGES_REQUESTED** [review]

File: src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java (hunk 29 above)
Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java:177]: What's the error in particular? I don't have a Mac to reproduce on. It's likely to do with how blank directories to use root are being put together.

Comment (tyBouch) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java]: Switched to using File.separator and converting separators to it. Lemme know how it flies.

Comment (dougnoel) [reply src/main/java/com/dougnoel/sentinel/filemanagers/FileManager.java]: Fixed
